scene/3d/navigation.h

```diff
@@ -31,154 +31,45 @@
-			return (a.key == p_key.a.key) ? (b.key < p_key.b.key) : (a.key < p_key.a.key);
-		};
-
-		EdgeKey(const Point &p_a = Point(), const Point &p_b = Point()) :
-				a(p_a),
-				b(p_b) {
-			if (a.key > b.key) {
-				SWAP(a, b);
-			}
-		}
-	};
-
-	struct NavMesh;
-	struct Polygon;
-
-	struct ConnectionPending {
-
-		Polygon *polygon;
-		int edge;
-	};
-
-	struct Polygon {
-
-		struct Edge {
-			Point point;
-			Polygon *C; //connection
-			int C_edge;
-			List<ConnectionPending>::Element *P;
-			Edge() {
-				C = NULL;
-				C_edge = -1;
-				P = NULL;
-			}
-		};
-
-		Vector<Edge> edges;
-
-		Vector3 center;
-		Vector3 entry;
-
-		float distance;
-		int prev_edge;
-		bool clockwise;
-
-		NavMesh *owner;
-	};
-
-	struct Connection {
-
-		Polygon *A;
-		int A_edge;
-		Polygon *B;
-		int B_edge;
-
-		List<ConnectionPending> pending;
-
-		Connection() {
-			A = NULL;
-			B = NULL;
-			A_edge = -1;
-			B_edge = -1;
-		}
-	};
-
-	Map<EdgeKey, Connection> connections;
-
-	struct NavMesh {
-
-		Object *owner;
-		Transform xform;
-		bool linked;
-		Ref<NavigationMesh> navmesh;
-		List<Polygon> polygons;
-	};
-
-	_FORCE_INLINE_ Point _get_point(const Vector3 &p_pos) const {
-
-		int x = int(Math::floor(p_pos.x / cell_size));
-		int y = int(Math::floor(p_pos.y / cell_size));
-		int z = int(Math::floor(p_pos.z / cell_size));
-
-		Point p;
-		p.key = 0;
-		p.x = x;
-		p.y = y;
-		p.z = z;
-		return p;
-	}
-
-	_FORCE_INLINE_ Vector3 _get_vertex(const Point &p_point) const {
-
-		return Vector3(p_point.x, p_point.y, p_point.z) * cell_size;
-	}
-
-	void _navmesh_link(int p_id);
-	void _navmesh_unlink(int p_id);
-
-	float cell_size;
-	Map<int, NavMesh> navmesh_map;
-	int last_id;
+	RID map;
 
 	Vector3 up;
-	void _clip_path(Vector<Vector3> &path, Polygon *from_poly, const Vector3 &p_to_point, Polygon *p_to_poly);
+	real_t cell_size;
+	real_t edge_connection_margin;
 
 protected:
 	static void _bind_methods();
+	void _notification(int p_what);
 
 public:
+	RID get_rid() const {
+		return map;
+	}
+
 	void set_up_vector(const Vector3 &p_up);
 	Vector3 get_up_vector() const;
 
-	//API should be as dynamic as possible
-	int navmesh_add(const Ref<NavigationMesh> &p_mesh, const Transform &p_xform, Object *p_owner = NULL);
-	void navmesh_set_transform(int p_id, const Transform &p_xform);
-	void navmesh_remove(int p_id);
+	void set_cell_size(float p_cell_size);
+	float get_cell_size() const {
+		return cell_size;
+	}
+
+	void set_edge_connection_margin(float p_edge_connection_margin);
+	float get_edge_connection_margin() const {
+		return edge_connection_margin;
+	}
 
 	Vector<Vector3> get_simple_path(const Vector3 &p_start, const Vector3 &p_end, bool p_optimize = true);
-	Vector3 get_closest_point_to_segment(const Vector3 &p_from, const Vector3 &p_to, const bool &p_use_collision = false);
-	Vector3 get_closest_point(const Vector3 &p_point);
-	Vector3 get_closest_point_normal(const Vector3 &p_point);
-	Object *get_closest_point_owner(const Vector3 &p_point);
```

Comment:
Were these functions intended to be removed? It'll be pretty difficult to write 3D point-and-click navigation code without these.

Reply:
I may add these back on the agent; before this can you please explain the use case, so I know better how it are intended to be used?

Reply:
Personally, I use `get_closest_point_to_segment` to raycast from a camera (e.g. so a player can click where they want a character to go). This functionality is also present in the 3D navmesh demo if you'd like a working example: https://github.com/godotengine/godot-demo-projects/tree/master/3d/navmesh

Reply:
The idea is to use the physics server to get this info, for this reason I removed these, but add this feature may be useful. I'll add it, thanks.

Reply:
I think I was using that, too - converting 2D map to 3D vectors, and then finding which segment (road) I was closest to (when creating the map, so in a tool script, where physics is not available nor necessary)

Reply:
@AndreaCatania Thank you for taking the time to do that

Reply:
But how do you intend to use physics server for querying navmesh nformation?

Reply:
Fixed by this #36269